File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/library/compositions/adapter/CompositionsAdapter.java
package com.github.anrimian.musicplayer.ui.library.compositions.adapter;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.utils.CompositionHelper;
import com.github.anrimian.musicplayer.ui.utils.OnPositionItemClickListener;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.diff_utils.SimpleDiffItemCallback;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.diff_utils.adapter.DiffListAdapter;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.github.anrimian.musicplayer.domain.Payloads.ITEM_SELECTED;
import static com.github.anrimian.musicplayer.domain.Payloads.ITEM_UNSELECTED;

/**
 * Created on 31.10.2017.
 */

public class CompositionsAdapter extends DiffListAdapter<Composition, MusicViewHolder> {

    private final Set<MusicViewHolder> viewHolders = new HashSet<>();

    private final HashSet<Composition> selectedCompositions;
    private final OnPositionItemClickListener<Composition> onCompositionClickListener;
    private final OnPositionItemClickListener<Composition> onLongClickListener;
    private final OnPositionItemClickListener<Composition> iconClickListener;

    @Nullable
    private Composition currentComposition;
    private boolean play;
    private boolean isCoversEnabled;

    public CompositionsAdapter(RecyclerView recyclerView,
                               HashSet<Composition> selectedCompositions,
                               OnPositionItemClickListener<Composition> onCompositionClickListener,
                               OnPositionItemClickListener<Composition> onLongClickListener,
                               OnPositionItemClickListener<Composition> iconClickListener) {
        super(recyclerView, new SimpleDiffItemCallback<>(
                CompositionHelper::areSourcesTheSame,
                CompositionHelper::getChangePayload)
        );
        this.selectedCompositions = selectedCompositions;
        this.onCompositionClickListener = onCompositionClickListener;
        this.onLongClickListener = onLongClickListener;
        this.iconClickListener = iconClickListener;
    }

    @NonNull
    @Override
    public MusicViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new MusicViewHolder(parent,
                onCompositionClickListener,
                onLongClickListener,
                iconClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull MusicViewHolder holder, int position) {
        viewHolders.add(holder);

        Composition composition = getItem(position);
        holder.bind(composition, isCoversEnabled);
        boolean selected = selectedCompositions.contains(composition);
        holder.setSelected(selected);

        boolean isCurrentComposition = composition.equals(currentComposition);
        holder.showAsCurrentComposition(isCurrentComposition);
        holder.showAsPlaying(isCurrentComposition && play);
    }

    @Override
    public void onBindViewHolder(@NonNull MusicViewHolder holder,
                                 int position,
                                 @NonNull List<Object> payloads) {
        if (payloads.isEmpty()) {
            super.onBindViewHolder(holder, position, payloads);
            return;
        }
        for (Object payload: payloads) {
            if (payload == ITEM_SELECTED) {
                holder.setSelected(true);
                return;
            }
            if (payload == ITEM_UNSELECTED) {
                holder.setSelected(false);
                return;
            }
        }
        holder.update(getItem(position), payloads);
    }

    @Override
    public void onViewRecycled(@NonNull MusicViewHolder holder) {
        super.onViewRecycled(holder);
        viewHolders.remove(holder);
    }

    public void setItemSelected(int position) {
        notifyItemChanged(position, ITEM_SELECTED);
    }

    public void setItemUnselected(int position) {
        notifyItemChanged(position, ITEM_UNSELECTED);
    }

    public void setItemsSelected(boolean selected) {
        for (MusicViewHolder holder: viewHolders) {
            holder.setSelected(selected);
        }
    }

    public void showCurrentComposition(Composition currentComposition) {
        this.currentComposition = currentComposition;
        for (MusicViewHolder holder: viewHolders) {
            Composition composition = holder.getComposition();
            boolean isCurrentComposition = composition.equals(currentComposition);
            holder.showAsCurrentComposition(isCurrentComposition);
            holder.showAsPlaying(isCurrentComposition && play);
        }
    }

    public void setCoversEnabled(boolean isCoversEnabled) {
        this.isCoversEnabled = isCoversEnabled;
        for (MusicViewHolder holder: viewHolders) {
            holder.setCoversVisible(isCoversEnabled);
        }
    }

    public void showPlaying(boolean play) {
        this.play = play;
        for (MusicViewHolder holder: viewHolders) {
            Composition composition = holder.getComposition();
            boolean isCurrentComposition = composition.equals(currentComposition);
            holder.showAsCurrentComposition(isCurrentComposition);
            holder.showAsPlaying(isCurrentComposition && play);
        }
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/library/compositions/adapter/CompositionsAdapter.kt
package com.github.anrimian.musicplayer.ui.library.compositions.adapter

import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.github.anrimian.filesync.models.state.file.FileSyncState
import com.github.anrimian.musicplayer.domain.Payloads
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.composition.CurrentComposition
import com.github.anrimian.musicplayer.domain.models.utils.CompositionHelper
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.diff_utils.SimpleDiffItemCallback
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.mvp.MvpDiffAdapter

/**
 * Created on 31.10.2017.
 */
open class CompositionsAdapter<T : Composition>(
    lifecycleOwner: LifecycleOwner,
    recyclerView: RecyclerView,
    private val selectedCompositions: Set<T>,
    private val onCompositionClickListener: (Int, T) -> Unit,
    private val onLongClickListener: (Int, T) -> Unit,
    private val iconClickListener: (Int, T) -> Unit,
    private val menuClickListener: (View, Int, T) -> Unit,
    diffCallback: DiffUtil.ItemCallback<T> = SimpleDiffItemCallback(
        CompositionHelper::areSourcesTheSame,
        CompositionHelper::getChangePayload
    )
) : MvpDiffAdapter<T, CompositionViewHolder<T>>(
    lifecycleOwner,
    recyclerView,
    diffCallback
) {

    private var currentComposition: CurrentComposition? = null
    private var isCoversEnabled = false
    private var syncStates = emptyMap<Long, FileSyncState>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CompositionViewHolder<T> {
        return CompositionViewHolder(
            parent,
            onCompositionClickListener,
            onLongClickListener,
            iconClickListener,
            menuClickListener
        )
    }

    override fun onBindViewHolder(holder: CompositionViewHolder<T>, position: Int) {
        super.onBindViewHolder(holder, position)

        val composition = getItem(position)
        holder.bind(composition, isCoversEnabled)

        val selected = selectedCompositions.contains(composition)
        holder.setSelected(selected)
        holder.showCurrentComposition(currentComposition, false)
        holder.setFileSyncStates(syncStates)
    }

    override fun onBindViewHolder(
        holder: CompositionViewHolder<T>,
        position: Int,
        payloads: List<Any>
    ) {
        if (payloads.isEmpty()) {
            super.onBindViewHolder(holder, position, payloads)
            return
        }
        for (payload in payloads) {
            if (payload === Payloads.ITEM_SELECTED) {
                holder.setSelected(true)
                return
            }
            if (payload === Payloads.ITEM_UNSELECTED) {
                holder.setSelected(false)
                return
            }
        }
        holder.update(getItem(position), payloads)
    }

    fun setItemSelected(position: Int) {
        notifyItemChanged(position, Payloads.ITEM_SELECTED)
    }

    fun setItemUnselected(position: Int) {
        notifyItemChanged(position, Payloads.ITEM_UNSELECTED)
    }

    fun setItemsSelected(selected: Boolean) {
        forEachHolder { holder -> holder.setSelected(selected) }
    }

    fun showCurrentComposition(currentComposition: CurrentComposition?) {
        this.currentComposition = currentComposition
        forEachHolder { holder ->
            holder.showCurrentComposition(currentComposition, true)
        }
    }

    fun setCoversEnabled(isCoversEnabled: Boolean) {
        this.isCoversEnabled = isCoversEnabled
        forEachHolder { holder ->
            holder.setCoversVisible(isCoversEnabled)
        }
    }

    fun showFileSyncStates(states: Map<Long, FileSyncState>) {
        this.syncStates = states
        forEachHolder { holder ->
            holder.setFileSyncStates(syncStates)
        }
    }
}